gms/gossiper.cc

```diff
@@ -1717,20 +1722,24 @@ void gossiper::examine_gossiper(utils::chunked_vector<gossip_digest>& g_digest_l
 }
 
 future<> gossiper::start_gossiping(int generation_number, bind_messaging_port do_bind) {
-    return start_gossiping(generation_number, std::map<application_state, versioned_value>(), do_bind);
+    return start_gossiping(generation_number, std::map<application_state, versioned_value>(), do_bind, advertise_myself_t::yes);
 }
 
-future<> gossiper::start_gossiping(int generation_nbr, std::map<application_state, versioned_value> preload_local_states, bind_messaging_port do_bind) {
+future<> gossiper::start_gossiping(int generation_nbr, std::map<application_state, versioned_value> preload_local_states, bind_messaging_port do_bind, advertise_myself_t advertise) {
     // Although gossiper runs on cpu0 only, we need to listen incoming gossip
     // message on all cpus and forard them to cpu0 to process.
-    return get_gossiper().invoke_on_all([do_bind] (gossiper& g) {
+    return get_gossiper().invoke_on_all([do_bind, advertise] (gossiper& g) {
         return g.init_messaging_service_handler(do_bind);
-    }).then([this, generation_nbr, preload_local_states] () mutable {
+    }).then([this, generation_nbr, preload_local_states, advertise] () mutable {
         build_seeds_list();
         if (_cfg.force_gossip_generation() > 0) {
             generation_nbr = _cfg.force_gossip_generation();
             logger.warn("Use the generation number provided by user: generation = {}", generation_nbr);
         }
+        if (!advertise) {
+            _initial_generation_number = generation_nbr;
```

Comment:
This needs a comment explaining what's going on.

Reply:
Also, when are we using `force_gossip_generation`?
It feels dangerous to me in conjunction with this change, can we get rid of it?

Reply:
Should we assert that `generation_nbr != 0` here?

Reply:
The force_gossip_generation is introduced (743b529c2b035658d33afd14f6a01beff3097b12) to help user fix cluster without patch 0a52ecb6df3deec729759ea87b4986334f207c07. We need some more time to get rid of this option.